status-im/status-react
# Account selection: Implement features related to editing shared addresses

- Fixes https://github.com/status-im/status-mobile/issues/18844
- Fixes https://github.com/status-im/status-mobile/issues/18931
- Fixes https://github.com/status-im/status-mobile/issues/18975
- Fixes https://github.com/status-im/status-mobile/issues/18835

### Summary

This larger than usual PR attempts to implement all flows for editing shared addresses (see [Figma](https://www.figma.com/file/h9wo4GipgZURbqqr1vShFN/Communities-for-Mobile?type=design&node-id=15340-270835&mode=design&t=gEgvqCtOPBL3BDM3-0)).

- Joining community flows should behave just as in `develop`.
- Editing addresses is only allowed behind a feature flag, which you can enable in `Profile > Feature Flags > edit-account-selection`. The flag is disabled by default. The goal is to enable or remove it once QAs deem the feature stable enough, thus not in this PR.

Note: I apologize for such a large diff to review. It's very unusual on my part and I don't endorse such practice. There are two reasons for this exception: the main one is that I'll be completely off for a month, so no PRs from me for a while ;) The other is that, while working for this PR, I realized many refactors were needed and they had to be consistently done, which would be difficult with multiple PRs and would require multiple QA passes.

### A few other fixes/improvements

- [x] We do less in event `:communities/handle-community`, which can be triggered numerous times. Overly dispatching there can be the cause of unwanted slow downs due to the amount of time spent computing derived data (e.g. getting revealed accounts) unnecessarily. As aligned with @cammellos, it's better if we call certain RPC endpoints only when really needed (e.g. in the mount phase of a component).
- [x] I fixed most of https://github.com/status-im/status-mobile/issues/18844, but not all. The issue can probably be closed. We still store in the community instance the _revealed accounts_ data. The technique to fix is repeatable, i.e. store in the app-db whatever data is needed, indexed by the community ID in a new root key.
- [x] In the screen to select addresses to share, local state is used, and the global state is only mutated once the user confirms. This helped solve issue https://github.com/status-im/status-mobile/issues/18835.
- [x] In account selection screens, fully migrated away from Reagent atoms.
- [x] Added a temporary global state for checking permissions, so that we can safely find the highest permission role while the user hasn't confirmed. This helped solve https://github.com/status-im/status-mobile/issues/18835.
- [x] We were using the `:checking?` state to decide if the confirmation button should be disabled while the request to check permissions to status-go had not finished. But because the time it takes for status-go to check permissions is very fast, for the user, it meant they would see the confirmation button flash on screen and things could jump around as well. The `:checking?` state is useful under the hood in the event handlers, but shouldn't be used in the UI because, generally speaking, loading indicators shouldn't be used for things that finish too quickly because they will be perceived as flashes on screen. They will also **negatively affect the performance of the initial mount phase** because there will be 3 render passes (off by default, then on, then off). Solves issue https://github.com/status-im/status-mobile/issues/18931
- [x] I tried to move events out of `communities/events.cljs` file that were more strongly related to account selection. I also moved account selection subs out of `status-im.subs.communities`. I think it's time we follow the steps of the wallet team and how they are breaking apart events, subs, data-store, schemas, etc.
- [x] Instead of dispatching 3 async requests to status-go to join a community or to edit shared addresses, I used a new approach in this repo where the individual RPC calls are promisified and then a single re-frame effect is used. With the help of promesa macros, things are easier to reuse and hopefully read as well. A similar strategy will eventually be used in contract tests.

### User flows

The following is a list of user flows I tested while working on this PR.

- User tries to join communities with/without various permissions just to see if there was any regression.
- While editing shared addresses, users can:
  - Toggle "Share all current and future addresses". This is only partially implemented because the choice is not persisted in the backend (related issue https://github.com/status-im/status-mobile/issues/18976).
  - See relevant tokens.
  - Select/deselect addresses to share and see the highest permission role text change on the fly.
  - If the user makes changes to the address selection and press the back button, they are asked if they want to discard changes before proceeding.
  - If the request to join is pending approval, and if the user deselects all addresses to share, then they are immediately given the option/button to "Cancel request". If canceled, they are redirected to the home screen.
  - If the user has already joined the community, and if the user deselects all addresses to share, then they are immediately given the option/button to "Leave community". If canceled, they are redirected to the home screen.
  - If the addresses to share do not include the previously selected airdrop address, the first available (selected) address will be used automatically for airdrops.

### Important

When the user changes to a lower role while selecting/deselecting addresses, the "Confirm changes" button should be displayed, but going to a higher role, a slider and authentication mechanism should be used (see in [Figma](https://www.figma.com/file/h9wo4GipgZURbqqr1vShFN/Communities-for-Mobile?type=design&node-id=16113-149606&mode=design&t=XCiBLOkxsecZRTdL-0)).

We can't fulfill this requirement right now because every single call to status-go to update addresses (or set the airdrop address) require signing some data, thus requiring some form of authentication. This point is not critical, but should be brought to designers' attention (for the record I talked with Mario about this, but it was before I was certain of how status-go was operating).

### Areas that may be impacted

In theory, nothing should be impacted. As long as the community joining flows behave the same, editing is behind a feature flag and should cause no harm.

### Steps to test

1. Log in and enable the feature flag `Profile > Feature Flags > edit-account-selection`.
2. Try to join any community and the account selection flows will start. No regressions should happen here.
3. For editing, go to the community overview screen on a joined or pending community, open the drawer and press on "Edit shared addresses".

The most important part to test from a QA perspective is if there are no regressions while trying to select addresses to join a community. Editing is less critical to get 100% right because it's behind a disabled feature flag.

Besides that, for testing the features related to edit shared addresses, I suggest going over the entirety of the page [Figma > Manage shared addresses](https://www.figma.com/file/h9wo4GipgZURbqqr1vShFN/Communities-for-Mobile?type=design&node-id=15340-270835&mode=design&t=gEgvqCtOPBL3BDM3-0).

Since this is a large unit of work and blocks other devs from collaborating in other PRs without massive rebase issues, we can try to fix only the most critical issues in  this PR, but if other, less important ones are found we should probably log separately.

status: ready
